designs/node-upgrades.md

```diff
@@ -0,0 +1,185 @@
+# Karpenter Node Upgrades - Automatic Upgrades
+
+## Overview
+
+Karpenter users are requesting for Node Upgrades, generally asking for [more complex node control](https://github.com/aws/karpenter/issues/1738) (91 up-votes and counting). While early versions of Karpenter meant for Provisioners to dictate node provisioning requirements and simple termination methods, Provisioners are becoming more semantically aligned with Node Groups. This is seen with [Provisioners owning nodes](https://github.com/aws/karpenter/pull/1934) and [Consolidation](https://github.com/aws/karpenter/pull/2123), where Karpenter controls the cluster in more ways than provisioning nodes for pending pods. To further align the Provisioner to represent a desired state of both new and existing nodes, Karpenter will implement [Node AMI Upgrades](https://github.com/aws/karpenter/issues/1716) by maintaining drifted nodes.
+
+The upgrade mechanism will also be scalable to other methods of spec drift in the future. We will answer the following questions:
+
+* How will we upgrade nodes?
+* Which nodes should Karpenter upgrade?
+* What can interfere with upgrades?
+
+If this does not prove enough, we can implement Imperative Upgrades or Maintenance Windows in the future.
+
+## User Stories
+
+* Karpenter will be able to automatically upgrade node AMIs to match their Provisioner requirements.
+* Karpenter will be able to enable other providers to implement their own drift mechanisms.
+
+### Background - What is node drift?
+
+Drifted nodes are nodes whose spec does not match the ProvisionerSpec and ProviderRefSpec. Automatic upgrades will monitor the cluster for drifted nodes and reconcile that node when necessary. A node can become drifted when a user changes their Provisioner or ProviderRef. For example, if a user who changes their architecture requirements to only `amd64` from `arm64`, any existing `arm64` nodes would be drifted. In another example, if a user changes their tags in the  will have underlying instances with “out-of-spec” tags. These are both actions done within the cluster.
+
+Yet, a node can drift even if the Provisioner or ProviderRef do not change. Underlying infrastructure in the Cloud Provider can be changed outside of the cluster. More specifically, using anything except the `aws-ids` filter in the `AWSNodeTemplate` `AMISelector` allows the AMI to change when a new AL2 EKS Optimized AMI is released, making nodes drift. Additionally, if a user changes the tags on their AMIs and uses arbitrary tags in the AWSNodeTemplate, nodes can also become drifted.
+
+## How will we upgrade nodes?
+
+When upgrading a node, Karpenter will minimize the downtime of the applications on the node by first creating the replacement node before terminating the old node. Once the replacement is ready, Karpenter will cordon and drain the old node, terminating it when it’s fully drained. Once the old node has been terminated, the upgrade will be complete.
```

Comment:
Suggest either "by creating the replacement node before terminating the old node" or "by first creating the replacement node then terminating the old node".